### DISCUSSION TITLE
PIA: automatically update forwarded port in qBittorrent or Deluge with shell + inotifywait

### DISCUSSION THREAD

**Discussion by hugalafutro** [qdm12/gluetun#809] [Ideas]
Assuming you have something like this in your gluetun `docker-compose.yml`:
```
    volumes:
      - ./data/port_forward:/tmp/gluetun/forwarded_port
```

`nano set_port.sh && chmod +x set_port.sh` alongside the docker-compose.yml:

for qBittorent (change `torrent`, `/path/to/` and `8532` according to your system):
```
#!/bin/sh
PIA_PORT=$(cat /path/to/data/port_forward)
docker exec torrent curl --data 'json={"listen_port": ${PIA_PORT}}' http://127.0.0.1:8532/api/v2/app/setPreferences
```

for deluge (change `torrent` and `/path/to/` according to your system):
```
#!/bin/sh
PIA_PORT=$(cat /path/to/data/port_forward)
docker exec torrent deluge-console -c /config "config --set listen_ports ($PIA_PORT,$PIA_PORT)"
```

(if you don't have inotifywait do `sudo apt install inotify-tools`
`crontab -e`
make a new line (change `/path/to` according to your system)
```
@reboot while inotifywait /path/to/data/port_forward;do /path/to/set_port.sh;done >/dev/null 2>&1
```
This will start the monitoring after reboot. To start the monitoring immediately without having to reboot do this (I think) :
```
while inotifywait /path/to/data/port_forward;do /path/to/set_port.sh;done & disown
``` 
when the file changes the script will read the new port number and run the script to change it in torrent client

**Comment by drunk-moe** [comment]
this is a couple of years old but I was looking for this - PIA seems to be fine with holding on to your ports for a long time but it seems protonvpn changes my ports frequently (multiple times a day). until now, I've just been manually changing it occasionally but having an automated way to do this would be nice! glad that I was able to come across this on google; I know this is an old post but I'll update this comment when I get around to testing it since others might come here from google as well

**Reply by hugalafutro** [reply]
I don't use this anymore as I switched providers, I went with a full machine routed via mullvad on which I run my "arr matey" stuff. Mullvad recently cancelled port forwarding so the point is moot now. As far as I remember if you try to use it exactly as I wrote it it will only work once, I messed up the loop or something, but I had it in a cron where I messed the @reboot so it ran each hour or something so I didn't notice. Strange thing though is when I torrent via mullvad, certain countries I am active peer even without port forwarding (works really well with Sweden servers for me).

Hope you're successful adapting it to your needs.

**Reply by drunk-moe** [reply]
I wrote a script! It's pretty basic and bad... but, it works! I'll post it as a response in case anyone needs it from Google hahaha. This helped point me in the right direction, though :)

**Comment by drunk-moe** [comment]
This works for me -- I run it via cron and have it output to a log file (for some reason it seems ProtonVPN servers go down a lot! Not sure if I can pass a list into Gluetun but that's unrelated) (also please don't judge me too harshly I threw this together in 15 minutes and it seems to work fine): 

my container names are `deluge` and `gluetun` so if you see that anywhere, you can change it to whatever your container name is.

```
#!/bin/bash
portfile='/path/to/forwarded_port_file'
dockstatus=$(docker ps)
curdate=$(date "+%Y-%m-%dT%H:%M:%S")
if echo $dockstatus | grep -q gluetun && echo $dockstatus | grep -q deluge; then
    if echo $dockstatus | grep gluetun | grep -v unhealthy | grep -q healthy; then
        if [ -f $portfile ]; then
            setport=$(cat $portfile)
            currentport=$(docker exec deluge deluge-console -c /config "config listen_ports" 2>/dev/null | grep listen_ports | sed -n 's/listen_ports: (\([0-9]*\).*/\1/p')
        else
            echo "$curdate - ERROR - port file not found!"
            exit 3
        fi
    else
        echo "$curdate - ERROR - gluetun container is unhealthy!"
        exit 2
    fi
else
    echo "$curdate - ERROR - deluge and/or gluetun not running!"
    exit 1
fi

if [ "$setport" -eq "$currentport" ]; then
    echo "$curdate - INFO - ports match, nothing to do"
    exit 0
else
    docker exec deluge deluge-console -c /config "config --set listen_ports ($setport,$setport)" >/dev/null 2>&1
    echo "$curdate - INFO - updated port to $setport!"
    exit 0
fi
```
